### DISCUSSION TITLE
How do you personally manage your stages/environments?

### DISCUSSION THREAD

**Discussion by bryanowen7** [serverless/serverless#10278] [Q&A] [answered]
When having multiple environments for developing, testing, QA and production releases, how do you/your team manage to keep them separate?

I've been developing with [serverless offline](https://www.npmjs.com/package/serverless-offline) and having two stacks, one for production and one for testing, but I find this kind of unorganized since the stack creates one function, API (Gateway), and log group per stage, so resources get "duplicated". This can get annoying if there is a large number of functions and/or stages.

Previously, I used function qualifiers with the [serverless-aws-alias](https://www.npmjs.com/package/serverless-aws-alias) plugin. That way, I was able to have a single function and log group, but have stages separated by version number. The API also had stages which pointed to each function's corresponding stage. However, this plugin has been long abandoned and is not compatible with Serverless v2.
I also want to avoid having to constantly modify the serverless.yml file to keep the CI/CD process as clean as possible.

Does anyone knows of any better way to manage this? I'm trying to keep my functions and log groups ordered the best way I can.

Thank you in advance.

**Comment by mnapoli** [comment, marked as answer]
Duplicating the whole stack and resources for each environment (called stage with Serverless Framework) is a good approach. It allows to keep everything separated: logs (which is a good thing), scalability limits (like "reserved concurrency" on Lambda functions), permissions, etc.

AWS even recommends using different AWS accounts to isolate production from the rest.
